### PR TITLE
Add AuctionFinalize trigger

### PR DESCRIPTION
**Pull request by gaborh-da**

### CODE REVIEW

File: src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        matchingBidCids =
+            getMatchingBidCids trigger.auctionAgent trigger.auctionName bidContracts
+    dedupExercise triggerCid AuctionFinalizeBotTrigger_AllocateBond with
+        participationCids = matchingParticipationCids
+        bidCids = matchingBidCids
+  where
+    getMatchingParticipationCids auctionAgent auctionName participationContracts =
+      map fst $
+        filter (\(_, p) -> p.auctionName == auctionName && p.auctionAgent == auctionAgent)
+          participationContracts
```
Comment (richardkapolnai-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml:58]: just wondering, what do you think about this?
```suggestion
      [ cid | (cid, p) <- participationContracts, p.auctionName == auctionName, p.auctionAgent == auctionAgent ]
```

File: src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+  finalizeTrigger <-
+    auctionAgent `submit`
+      exercise auction Auction_Finalize
+
+  let activeContracts =
+          toACS participation1 <> toACS participation2 <>
+          toACS bid1 <> toACS finalizeTrigger
+  commands <-
+    testRule auctionFinalizeTrigger auctionAgent activeContracts DA.Next.Map.empty ()
```
Comment (chrisnorris-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml:109]: maybe use M.empty

File: src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+  finalizeTrigger <-
+    auctionAgent `submit`
+      exercise auction Auction_Finalize
+
+  let activeContracts =
+          toACS participation1 <> toACS participation2 <>
+          toACS bid1 <> toACS bid2 <> toACS finalizeTrigger
+  commands <-
+    testRule auctionFinalizeTrigger auctionAgent activeContracts DA.Next.Map.empty ()
```
Comment (chrisnorris-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml:147]: M.empty

File: src/main/daml/DA/RefApps/Bond/Triggers/CommissionTrigger.daml [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  let activeContracts = toACS issuance.bondBundleData.assetDepositCid <>
+                        toACS issuance.bondBundleData.bondSettlementCid <>
+                        toACS issuance.bondBundleData.assetFungibleCid <>
+                        toACS issuance.bondBundleData.cashSettlementCid <>
+                        toACS issuance.fixedRateBondFactCid
+  commands <-
+    testRule commissionTrigger issuer activeContracts DA.Next.Map.empty ()
+
+  let flattenedCommands = flattenCommands commands
+  length flattenedCommands === 0
```
Comment (chrisnorris-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/CommissionTrigger.daml:138]: ```(testRule commissionTrigger issuer activeContracts M.empty ()) >>= (( === 0 ). length)```

File: src/main/daml/DA/RefApps/Bond/Triggers/CommissionTrigger.daml [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    commissionAuction issuer issuance
+
+  let activeContracts = toACS commissionBotTrigger <>
+                        toACS issuance.bondBundleData.assetDepositCid <>
+                        toACS issuance.bondBundleData.bondSettlementCid <>
+                        toACS issuance.bondBundleData.assetFungibleCid <>
+                        toACS issuance.bondBundleData.cashSettlementCid <>
+                        toACS issuance.fixedRateBondFactCid
+  commands <-
+    testRule commissionTrigger issuer activeContracts DA.Next.Map.empty ()
```
Comment (chrisnorris-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/CommissionTrigger.daml:108]: again perhaps ```M.empty```?

File: src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+module DA.RefApps.Bond.Triggers.AuctionFinalizeTrigger where
+
+import Daml.Trigger
+import DA.Next.Map qualified as M
+import DA.Foldable hiding (length)
+import Daml.Trigger.Assert
+import DA.Assert
+import DA.RefApps.Bond.Test.FullScenario qualified as MarketSetup
+import DA.RefApps.Bond.Auction
+import DA.Next.Map qualified
```
Comment (chrisnorris-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/AuctionFinalizeTrigger.daml:17]: can remove

File: src/main/daml/DA/RefApps/Bond/Triggers/CommissionTrigger.daml [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+module DA.RefApps.Bond.Triggers.CommissionTrigger where
+
+import Daml.Trigger
+import DA.Next.Map qualified as M
+import DA.Foldable hiding (length)
+import Daml.Trigger.Assert
+import DA.Assert
+import DA.RefApps.Bond.Test.FullScenario qualified as MarketSetup
+import DA.Next.Map qualified
```
Comment (chrisnorris-da) [inline src/main/daml/DA/RefApps/Bond/Triggers/CommissionTrigger.daml:16]: can remove

File: src/main/daml/DA/RefApps/Bond/ContractUtils.daml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+headNote: Text -> [a] -> a
+headNote message = (fromSomeNote message) . head
+
+findAsset: (HasAccount a) => Party -> Party -> [Contract a] -> ContractId a
+findAsset provider owner assets =
+  headNote "Can't find any asset contract" $ filterContractIds (accountBelongsTo provider owner) assets
+
+accountBelongsTo: (HasAccount a) => Party -> Party -> a -> Bool
+accountBelongsTo provider owner a = do
+  let account = accountOf a
```
Comment (gaborh-da) [inline src/main/daml/DA/RefApps/Bond/ContractUtils.daml:40]: one can use `let ... in ...` without do

**Pull request merged**